Import level two licence sectors into Specialist Publisher

leenagupte opened the pull request:
Trello: https://trello.com/c/TiYLM2hE

# What's changed and why?

It has been agreed with Content Designers and User Researchers to tag migrated licences in Specialist Publisher to only level two sectors (from Licence Finder).

The hierarchical structure for sectors (3 levels) is being dropped and licence documents will just be tagged to the 2nd level. This is due to limitations to support hierarchical tagging in Specialist Publisher / Finder Frontend and to mitigate bespoke work just for User Research.

Activities do not need to be imported as activities don't make complete sense. They mostly map 1-2-1 with individual licences but have different names to the licences.

Country tagging will remain and is already configured in Specialist Publisher.

The tagging of licence documents to the level sectors will be accomplished in a future data migration.

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

leenagupte committed:
Text file with level 2 sectors from Licence Finder

File was created by running the following query against the
Licence Finder mongo database and copying the output to file:

```
Sector.where(layer: 2).pluck(:name).sort
```

This file will only be used to import the industry sector facets into
the new licence transaction specialist document. Once the new specialist
finder has gone live this file will be removed.

In the original Licence Finder, industry sectors are nested three levels
deep. Specialist Finders don't have a component that supports this type
of nesting, so it has been decided to only import the level 2 sectors.

The sector names will be modified in a later change to match the
language used by the users.

leenagupte committed:
Import industry sectors from file

This class will be used to populate the industry sector facets in the
licence transaction schema.

leenagupte committed:
Update sector values in licence transaction schema

JSON was generated by running:

```
 bundle exec rake licence_transaction:import_industry_sectors
```

`JSON.dump` outputs the JSON file on to one line which is not very
readable. The editor's (in this case VS Code) "format document" function
has been used to prettify the code, which is why the formatting in the
rest of the schema has changed.

1pretz1 reviewed (approved):
Looks good!

leenagupte committed:
Write the updated JSON directly to the schema file

Writes the industry sectors directly to the JSON file for the
licence transaction schema.

1pretz1 commented on app/services/licence_transaction/import_industry_sectors.rb:
```diff
@@ -4,6 +4,10 @@ def self.call
   end
```
Minor: we should use encapsulation for the private methods

1pretz1 commented on spec/services/licence_transaction/import_industry_sectors_spec.rb:
```diff
@@ -1,21 +1,21 @@
 require "spec_helper"
 
 RSpec.describe LicenceTransaction::ImportIndustrySectors do
```
I think we're now testing the wrong method (should be `call`) and as a result we're missing some tests for the features that have been added (reading / writing to JSON file). If we find it useful to have `imported_json_data` as its own public method then we should probably test both.

leenagupte replied:
You caught me! I'll have a bit of refactor and find a nice way to test the JSON file.

1pretz1 reviewed (commented):
Changes look good, nice automation! I've just made a couple of comments. We could also create a rake task for this import which is a bit nicer than having to run commands in a console.

leenagupte committed:
Pass file paths to BulkIndustrySectorsImporter

This allows the file path methods to be made private, and to use fixture
data in the tests.
If no file paths are passed in, defaults are used.

leenagupte committed:
Add test for writing updated sectors to JSON file

leenagupte committed:
Make sector import methods private

The only method that should be called outside of the class is `call`.
The other methods should be private.

Also combines the tests to only test the call method.

leenagupte committed:
Add a rake task to import industry sectors

1pretz1 commented on app/services/licence_transaction/import_industry_sectors.rb:
```diff
@@ -0,0 +1,43 @@
+class LicenceTransaction::ImportIndustrySectors
```
This feels like it should live in lib as it's only used by a single rake task and not part of the main application code

leenagupte replied:
Hmmm, I thought about this, and figured it was really an import service with the potential to be run a few times. Maybe even refactored to be used in the scripts to update the sector names and re-tag licences. 

The name is probably wrong though. I think it should have been named "IndustrySectorsImporter"

1pretz1 replied:
I still think it should live in `lib` unless it is being used by the application at runtime as everything else in the services directory is related to the main application code. Interestingly, we do have an `importers` directory in `lib`: https://github.com/alphagov/specialist-publisher/tree/589ad0fee62548e86e979aec74841044e841f959/lib.

Agreed on the proposed name 👍

leenagupte replied:
Agreed on name and location. I'll move the class and test files into lib. I noticed the same "importer" folder when I was reviewing your PR! https://github.com/alphagov/specialist-publisher/pull/2210#discussion_r1072103633

1pretz1 reviewed (approved):
Nice work! Approving as my only comment is minor.

1pretz1 reviewed (approved):
Nice!
